CIScripts/Test/PesterHelpers/PesterHelpers.ps1

```diff
@@ -51,5 +51,31 @@ function Eventually {
         [Parameter(Mandatory=$false)] [int] $Interval = 1,
         [Parameter(Mandatory=$true)] [int] $Duration = 3
     )
-    $ScriptBlock | Invoke-UntilSucceeds -Interval $Interval -Duration $Duration
+    if ($Duration -lt $Interval) {
+        throw "Duration must be longer than interval"
+    }
+    if ($Interval -eq 0) {
+        throw "Interval must not be equal to zero"
+    }
+    $StartTime = Get-Date
+    $DidNotThrow = $false
+    do {
+        try {
+            & $ScriptBlock
+            $DidNotThrow = $true
+        } catch {
+            $DidNotThrow = $false
+            $LastException = $_.Exception
+        } finally {
+            Start-Sleep -s $Interval
+        }
+        if($DidNotThrow) {
+            return
+        }
+    } while (((Get-Date) - $StartTime).Seconds -lt $Duration)
```

Comment:
Let's assume `interval` is 5, `duration` is 15 and `scriptBlock` executes for 10 seconds and throws an exception. Then `Eventually` will call `scriptBlock` only once, instead of 3 times.

For now let's use the following:

```
sum = 0
do {
    # do stuff
    sleep(interval)
    sum += interval
} while (sum < duration)
```

Reply:
I'd say, let's do:

```powershell
while ($true) {
    # stuff
    if (((Get-Date) - $StartTime).Seconds -gt $Duration) { break }
    start-sleep -s interval
}
```

With that approach, your 10-second scriptblock with 15-second timeout will twice. I'd say it's ok not to run three times, as we specify duration, not the iteration count, so the length of the test itself shouldn't artificially elongate the timeout.

Reply:
Actually, let's do:

```powershell
$ShouldRepeat = $true
while ($true) {
    $ShouldRepeat = (((Get-Date) - $StartTime).Seconds -le $Duration) { break }
    # stuff
    if (-not $ShouldRepeat) { break }
    start-sleep -s interval
}